File: inJava/app/src/main/java/com/barrygithub/rickandmortyapp/data/LocalDatasource/MyDAO.java
package com.barrygithub.rickandmortyapp.data.LocalDatasource;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import com.barrygithub.rickandmortyapp.data.LocalDatasource.entities.Character;
import com.barrygithub.rickandmortyapp.data.LocalDatasource.entities.EntityDb;
import com.barrygithub.rickandmortyapp.data.LocalDatasource.entities.MetaData;

import java.util.List;

import io.reactivex.rxjava3.core.Single;

/**
 * Project RickAndMortyApp
 * Created by Barry Zea H. on 5/04/23.
 * Copyright (c) Barry Zea H. All rights reserved.
 **/
@Dao
public interface MyDAO {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    Void saveMetadata(MetaData metadata);

    @Update
    Single<Integer> updateMetaData(MetaData metaData);

    @Insert
    Void saveCharacters(List<Character> characters);

    @Query("select * from MetaData")
    Single<List<MetaData>> getMetadata();


    @Query("select * from Character")
    Single<List<Character>> getCharacters();

    @Transaction
    @Query("select * from MetaData")
    Single<EntityDb> getDataFromDatabase();


}

File: inKotlin/app/src/main/java/com/barrygithub/rickandmortyapp/data/localDatasource/MyDAO.kt
package com.barrygithub.rickandmortyapp.data.localDatasource

import androidx.room.*
import io.reactivex.rxjava3.core.Single

/**
 * Project RickAndMortyApp
 * Created by Barry Zea H. on 14/04/23.
 * Copyright (c) Barry Zea H. All rights reserved.
 *
 **/
@Dao
interface MyDAO {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun saveMetadata(metadata: MetaData)

    @Insert
    fun saveCharacters(characters: List<Character>)

    @Query("select * from Character")
    fun getCharacters(): Single<List<Character>>

    //con este método traemos la metadata como la lista de personajes relacionados
    @Transaction
    @Query("select * from MetaData")
    fun getDataFromDatabase(): Single<EntityDb>
}